HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,16 @@
+import pyspark
+from pyspark import SparkContext, SparkConf
+#sc = SparkContext()
+
+def bfs(graph, starting_char):
+    char_list = [starting_char]
+    BFS_list  = []
+    
+    while len(char_list)!=0:
+        neighbors = graph.filter(lambda tpl: tpl[0] in char_list).flatMap(lambda tpl: tpl[1]).collect()
+        BFS_list  = BFS_list + char_list
+        char_list = []
+        char_list = char_list + list(set(neighbors)-set(BFS_list))                                     
+        print len(list(set(neighbors)-set(BFS_list)))
```

Comment:
Your implementation here doesn't take much advantage of parallelism. You should be able to do each round in the loop with one map and one reduce. One way to do this, is to make `graph` an [adjacency list](https://en.wikipedia.org/wiki/Adjacency_list) that also stores distance, and overwrite it each time, updating the distances of the nodes touched in that round. You can use an accumulator to figure out when to stop (this is what you'd check in the `while` condition), and the output would be an adjacency list of all characters and their distance from the source.

This takes better advantage of parallelism and reduces costs imposed by `collect()`, which is a very expensive operation.